File: frontend/src/lib/printBrand.ts
/** Shared Yash Gallery brand block for printable ERP documents. */

import { YASH_GALLERY_LOGO_DATA_URL } from './yashGalleryLogoDataUrl'

export const YASH_GALLERY = {
  name: 'Yash Gallery Pvt. Ltd.',
  shortName: 'Yash Gallery',
  address: '55 TO 64, Tatiyawas, Brij Vihar, Amber, Jaipur (Raj.) 303704',
  gst: '08AABCY3804E1ZJ',
  phone: '',
  email: 'purchase@yashgallery.com',
}

/** Absolute URL to /logo.png (works in app UI). */
export function brandLogoUrl(): string {
  if (typeof window === 'undefined') return '/logo.png'
  return `${window.location.origin}/logo.png`
}

/**
 * Logo as an embedded data URL (no network / Cloudflare / about:blank issues).
 * Sync-friendly: always available for print HTML builders.
 */
export async function brandLogoDataUrl(): Promise<string> {
  return YASH_GALLERY_LOGO_DATA_URL
}

/** Sync accessor for builders that already await other work. */
export function brandLogoDataUrlSync(): string {
  return YASH_GALLERY_LOGO_DATA_URL
}

export type BrandHeaderOpts = {
  department?: string
  docTitle: string
  docNumber: string
  barcodeHtml?: string
  /** Extra HTML under company meta (e.g. style thumb) */
  extraLeft?: string
  /** Logo as data URL or absolute URL */
  logoSrc?: string
}

/** Left company block + right document title/number. */
export function brandPrintHeaderHtml(opts: BrandHeaderOpts): string {
  const logo = opts.logoSrc || YASH_GALLERY_LOGO_DATA_URL
  const dept = opts.department || ''
  const barcode = opts.barcodeHtml || ''
  return `
    <div class="header">
      <div class="company-block">
        <img src="${logo}" alt="Yash Gallery" class="company-logo" />
        <div>
          <div class="company-name">${YASH_GALLERY.name}</div>
          <div class="company-meta company-address">${YASH_GALLERY.address.replace(/\n/g, '<br/>')}</div>
          ${YASH_GALLERY.gst ? `<div class="company-meta"><strong>GSTIN:</strong> ${YASH_GALLERY.gst}</div>` : ''}
          ${YASH_GALLERY.phone ? `<div class="company-meta">Tel: ${YASH_GALLERY.phone}</div>` : ''}
          ${YASH_GALLERY.email ? `<div class="company-meta">${YASH_GALLERY.email}</div>` : ''}
          ${dept ? `<div class="company-meta company-dept">${dept}</div>` : ''}
          ${opts.extraLeft || ''}
        </div>
      </div>
      <div>
        <div class="doc-title">${opts.docTitle}</div>
        <div class="doc-num">${opts.docNumber}</div>
        ${barcode ? `<div style="margin-top:8px;display:flex;justify-content:flex-end">${barcode}</div>` : ''}
      </div>
    </div>`
}

/** Shared CSS snippet for brand header (include in print window styles). */
export const BRAND_PRINT_CSS = `
    .header{display:flex;justify-content:space-between;align-items:flex-start;border-bottom:2px solid #002B5B;padding-bottom:12px;margin-bottom:16px;gap:16px}
    .company-block{display:flex;gap:14px;align-items:flex-start;min-width:0;flex:1}
    .company-logo{height:64px;width:auto;max-width:180px;object-fit:contain;display:block;flex-shrink:0}
    .company-name{font-size:18px;font-weight:700;color:#002B5B;line-height:1.25}
    .company-meta{font-size:11px;color:#334155;line-height:1.45;margin-top:3px}
    .company-address{font-size:11px;color:#1e293b;font-weight:500}
    .company-dept{margin-top:6px;font-weight:700;color:#002B5B}
    .doc-title{font-size:16px;font-weight:600;color:#002B5B;text-align:right}
    .doc-num{font-size:22px;font-weight:800;color:#002B5B;text-align:right}
`

File: backend/services/print_brand.py
"""Shared Yash Gallery brand HTML for server-rendered printable documents."""
from __future__ import annotations

import base64
from functools import lru_cache
from pathlib import Path

YASH_GALLERY = {
    "name": "Yash Gallery Pvt. Ltd.",
    "short_name": "Yash Gallery",
    "address": "Bhiwandi, Thane District, Maharashtra — 421302, India",
    "email": "purchase@yashgallery.com",
}

_LOGO_CANDIDATES = (
    Path(__file__).resolve().parents[1] / "static" / "logo.png",
    Path(__file__).resolve().parents[2] / "frontend" / "public" / "logo.png",
    Path("/app/frontend/public/logo.png"),
    Path("/root/app/frontend/public/logo.png"),
    Path("/root/app/backend/static/logo.png"),
)


@lru_cache(maxsize=1)
def logo_data_url() -> str:
    try:
        from .print_brand_logo import YASH_GALLERY_LOGO_DATA_URL

        if YASH_GALLERY_LOGO_DATA_URL:
            return YASH_GALLERY_LOGO_DATA_URL
    except Exception:
        pass
    for path in _LOGO_CANDIDATES:
        try:
            if path.is_file():
                raw = path.read_bytes()
                b64 = base64.b64encode(raw).decode("ascii")
                return f"data:image/png;base64,{b64}"
        except OSError:
            continue
    return "/logo.png"


def brand_header_html(
    *,
    doc_title: str,
    doc_number: str = "",
    department: str = "",
    barcode_html: str = "",
    extra_left: str = "",
) -> str:
    logo = logo_data_url()
    addr = YASH_GALLERY["address"]
    email = YASH_GALLERY.get("email") or ""
    dept = (
        f'<div style="margin-top:6px;font-size:11px;font-weight:700;color:#002B5B">{department}</div>'
        if department
        else ""
    )
    email_html = f'<div style="font-size:11px;color:#334155;margin-top:2px">{email}</div>' if email else ""
    num = (
        f'<div style="font-size:20px;font-weight:800;color:#002B5B;text-align:right">{doc_number}</div>'
        if doc_number
        else ""
    )
    bc = (
        f'<div style="margin-top:8px;display:flex;justify-content:flex-end">{barcode_html}</div>'
        if barcode_html
        else ""
    )
    return f"""
    <div class="header" style="display:flex;justify-content:space-between;align-items:flex-start;gap:16px;border-bottom:2px solid #002B5B;padding-bottom:12px;margin-bottom:16px">
      <div style="display:flex;gap:14px;align-items:flex-start;min-width:0;flex:1">
        <img src="{logo}" alt="Yash Gallery" style="height:64px;width:auto;max-width:180px;object-fit:contain;display:block;flex-shrink:0" />
        <div>
          <div style="font-size:18px;font-weight:700;color:#002B5B;line-height:1.25">{YASH_GALLERY["name"]}</div>
          <div style="font-size:11px;color:#1e293b;font-weight:500;line-height:1.45;margin-top:3px">{addr}</div>
          {email_html}
          {dept}
          {extra_left}
        </div>
      </div>
      <div>
        <div style="font-size:16px;font-weight:600;color:#002B5B;text-align:right">{doc_title}</div>
        {num}
        {bc}
      </div>
    </div>"""
